### DISCUSSION TITLE
Java build process

### DISCUSSION THREAD

**Discussion by adammetcalf** [OpenDDS/OpenDDS#4730] [Q&A] [answered]
Hello,

I have been working through the developers guide in advance of tackling a large project which by necessity requires use of Java.

The documentation for the c++ side of things is mostly very good, and I have been able to understand in good detail how to configure and build an example 'Hello World', which I have successfully deployed as a pair of standalone executables.

I'm now working through similar for java, and I have noticed that the documentation is far less comprehensive.

Just as an example, the Java documentation adds good detail about building the necessary files for the idl definition, much in the same way as was provided for c++. Where it begins to miss necessary details is the publisher, subscriber and the DataReaderListenerImpl. There are no build steps here. 

Is it as simple as running mpc to create the build files (in my case using visual studio) and then building the generated solution?

ie: 
Run MPC to generate platform-specific build files:
```
%ACE_ROOT%\bin\mwc.pl -type vs2019
```

Open Visual Studio scoped to the generated sln:
```
devenv HelloWorld_idl.sln
```

**Comment by mitza-oci** [comment]
MPC can be used as the build system for the Java applications.  This is done for OpenDDS's included Java tests (see `java/tests/messenger/...` for some examples).

However, most Java-centric applications will use their own Java-centric (or JVM-centric) build systems.  To do this, first build the IDL you need for the application using the steps documented (hopefully this is sufficiently documented, if not we'll fix it).  The outputs of this step are a single JAR file and a native library for your platform (or one per platform).  The Java-centric build system should support existing external JAR files as dependencies.

**Reply by adammetcalf** [reply]
Thanks. 

Just for clarification, the native library refers to the dll if I'm working in windows (and I guess .so if in Linux)? ie in the messenger example the jar file is messenger_idl_test.jar and the native library is messenger_idl_testd.dll?

I'm guessing that I must also include the jar files from OpenDDS_ROOT/lib into my eclipse project too. This reason I ask for this clarification is that when building the OpenDDS directory (fresh from downloading) all the supporting dlls have a 'd' tagged on the end, and this seems to break the linker.

As an example, in the OpenDDS_ROOT/lib there is OpenDDS_DCPS_Java**d**.dll, but when importing the jar files as external jar files and linking the directory for the supporting dll I get an error in eclipse when I try to run anything.  The error states that it cannot find OpenDDS_DCPS_Java.dll.

**Reply by mitza-oci** [reply, marked as answer]
> Just for clarification, the native library refers to the dll if I'm working in windows (and I guess .so if in Linux)? ie in the messenger example the jar file is messenger_idl_test.jar and the native library is messenger_idl_testd.dll?

That's correct, for a Windows build using the Debug configuration.

> I'm guessing that I must also include the jar files from OpenDDS_ROOT/lib into my eclipse project too. 

Yes, the artifacts generated by building OpenDDS (also both JARs and native libraries) are needed.

> This reason I ask for this clarification is that when building the OpenDDS directory (fresh from downloading) all the supporting dlls have a 'd' tagged on the end, and this seems to break the linker.
> 
> As an example, in the OpenDDS_ROOT/lib there is OpenDDS_DCPS_Java**d**.dll, but when importing the jar files as external jar files and linking the directory for the supporting dll I get an error in eclipse when I try to run anything. The error states that it cannot find OpenDDS_DCPS_Java.dll.

See step 7 of https://opendds.readthedocs.io/en/latest-release/devguide/java_bindings.html#setting-up-an-opendds-java-project which has a Windows-specific paragraph about how to set it up to use the `*d.dll` files.  In OpenDDS's included tests, this is handled by code in  `bin/PerlDDS/Process_Java.pm`.

**Reply by adammetcalf** [reply]
Nice one, thanks very much!
